### PR TITLE
WIP Elm 0.19

### PR DESCRIPTION
**Pull request by avh4**
## TODO before merging this
  - [x] read through the entire test run output and make sure there's nothing unexpected happening.  Test output can be seen via the Travis-CI build https://travis-ci.org/rtfeldman/node-test-runner/builds
  - [x] switch to elmi-to-json-0.19.0-rc3 (waiting on https://github.com/stoeffel/elmi-to-json/pull/11) to reliably find elmi-to-json binary

## TODO (between rc release and stable release)
  - [ ] Make sure formatting of `String.join ","` places are okay
  - [ ] restore the "(Elm 0.19 beta allows multiple ports with the same name?)" test after that gets fixed in Elm https://elmlang.slack.com/archives/C13L7S5GR/p1528239699000411
  - [ ] restore `.travis.yml` after official Elm binaries are published
  - [ ] error message "No tests found in the test/ (or tests/) directory." should only allow "tests"
  - [ ] remove binstall dependency after updating to elmi-to-json-0.19.0-rc4

### CODE REVIEW

File: installer.js [hunk 1]
```diff
@@ -2,6 +2,10 @@ module.exports = function(verbose) {
   var binstall = require("binstall");
   var path = require("path");
   var fs = require("fs");
+
+  fs.copyFileSync("/Users/avh4/workspace/elm19/bin/elmi-to-json-30d8b8cb", path.join(__dirname, "bin/elm-interface-to-json"));
+  return new Promise(function (resolve) { resolve(); });;
```
Comment (avh4) [inline installer.js]: TODO: publish elm-interface-to-json and remove this!

File: lib/elm-test.js [hunk 2]
```diff
@@ -350,18 +349,18 @@ function runElmTest() {
...
+      // // just downloaded and compiled.
+      // var newElmStuffPath = path.join(newElmPackageDir, "elm-stuff");
+
+      // if (!fs.existsSync(newElmStuffPath)) {
+      //   fs.symlinkSync(
+      //     path.join(testRootDir, "elm-stuff"),
+      //     newElmStuffPath,
+      //     "junction" // Only affects Windows, but necessary for this to work there. See https://github.com/gulpjs/vinyl-fs/issues/210
+      //   );
+      // }
```
Comment (avh4) [inline lib/elm-test.js]: The new elm-compiler doesn't seem to like symlinks or something; it seemed to be getting confused about what it needed to rebuild, and the tests would fail to compile.  However, I think just removing this (and not having generated-code share elm-stuff with the tests folder) should be fine because Elm 0.19 compiles so much faster.

File: lib/elm-test.js [hunk 3]
```diff
@@ -503,33 +502,9 @@ function generatePackageJson(filePathArgs) {
...
-              );
-            }
-            resolve();
-          }
-        );
-      }
-    });
-  });
+  // This is no longer necessary in Elm 0.19
+  // TODO: remove this function
```
Comment (avh4) [inline lib/elm-test.js]: TODO: do this

Comment (avh4) [reply lib/elm-test.js]: done in 794dd16

File: lib/runner.js [hunk 4]
```diff
@@ -159,7 +106,10 @@ function findTests(
 )/*:Promise<void>*/ {
   return new Promise(function(resolve, reject) {
     function finish() {
-      var process = spawn(readElmiPath, ["--path", elmPackageJsonPath]);
+      console.error(global.process.cwd());
+      console.error(elmPackageJsonPath);
+      // NOTE: must be cd'd into elmPackageJsonPath when running elm-interface-to-json
```
Comment (avh4) [inline lib/runner.js]: TODO: verify that this is the case

File: src/Test/Reporter/Console/Format.elm [hunk 5]
```diff
@@ -122,21 +127,21 @@ listDiffToString index description { expected, actual } originals =
             , "This should never happen!"
             , "Please report this bug to https://github.com/elm-community/elm-test/issues - and include these lists: "
             , "\n"
-            , toString originals.originalExpected
+            , String.join ", " originals.originalExpected
```
Comment (avh4) [inline src/Test/Reporter/Console/Format.elm]: TODO: is there a function somewhere that will nicely format lists for us?

File: tests/DependencyPassing.elm [hunk 6]
```diff
@@ -1,16 +1,16 @@
 module DependencyPassing exposing (..)
 
 import Expect
-import String.Extra
+import Regex
```
Comment (avh4) [inline tests/DependencyPassing.elm]: (Switched to Regex for this test because `elm/regex` was already published, and `elm-community/string-extra` isn't published yet and needs a bit of work to update for 0.19)

File: tests/ci.js [hunk 7]
```diff
@@ -132,14 +132,16 @@ ls("tests/*.elm").forEach(function(testToRun) {
   } else if (/Failing\.elm$/.test(testToRun)) {
     echo("\n### Testing " + testToRun + " (expecting it to fail)\n");
     assertTestFailure(testToRun);
-  } else if (/PortRuntimeException.elm$/.test(testToRun)) {
+  } else if (/PortRuntimeException\.elm$/.test(testToRun)) {
+    echo("\n### TODO " + testToRun + " (Elm 0.19 beta allows multiple ports with the same name?)");
```
Comment (avh4) [inline tests/ci.js]: TODO: figure out if this is a bug in Elm 0.19 beta

File: installer.js [hunk 8]
```diff
@@ -2,6 +2,10 @@ module.exports = function(verbose) {
   var binstall = require("binstall");
   var path = require("path");
   var fs = require("fs");
+
+  fs.copyFileSync("/Users/avh4/workspace/elm19/bin/elmi-to-json-a650892", path.join(__dirname, "bin/elm-interface-to-json"));
+  return new Promise(function (resolve) { resolve(); });;
```
Comment (avh4) [inline installer.js]: TODO: publish elm-interface-to-json, and remove this

File: .travis.yml [hunk 9]
```diff
@@ -26,14 +27,31 @@ before_install:
...
+      tar zxvf linux-64.tar.gz
+      shasum elm
+      curl -O https://a.uguu.se/EcO4p2DxVWZx.
+      echo "adad27f79f3489e2e0da8fb933cc3e70f662c15d  EcO4p2DxVWZx." | shasum --check
+      mv EcO4p2DxVWZx. elm-interface-to-json
+      chmod a+x elm-interface-to-json
+      cd ..
+    fi
+  - ls "$(pwd)/elm19"
+  - export PATH="$(pwd)/elm19:$PATH"
```
Comment (avh4) [inline .travis.yml]: All of the changes in this file can be reverted once the new versions of elm-interface-to-json and elm are published.

File: installer.js [hunk 10]
```diff
@@ -2,6 +2,10 @@ module.exports = function(verbose) {
   var binstall = require("binstall");
   var path = require("path");
   var fs = require("fs");
+
+  fs.copyFileSync("elm19/elm-interface-to-json", path.join(__dirname, "bin/elm-interface-to-json"));
+  return new Promise(function (resolve) { resolve(); });;
```
Comment (avh4) [inline installer.js]: TODO: publish elm-interface-to-json, and remove this

File: .gitignore [hunk 11]
```diff
@@ -28,5 +28,5 @@ bin/elm-interface-to-json
 # Dependency directory
 # https://www.npmjs.org/doc/misc/npm-faq.html#should-i-check-my-node_modules-folder-into-git
 node_modules
-elm-stuff
+/elm-stuff
```
Comment (eeue56) [inline .gitignore]: Is this change needed? don't we just want to ignore all elm-stuff?

File: lib/elm-test.js [hunk 12]
```diff
@@ -86,20 +83,17 @@ function resolveFilePath(filename) {
   });
 }
 
-var pathToMake = undefined;
-var pathToElmPackage = "elm-package";
+var pathToElm = "elm";
 
 if (args.compiler !== undefined) {
-  pathToMake = path.resolve(args.compiler);
+  pathToElm = path.resolve(args.compiler);
```
Comment (eeue56) [inline lib/elm-test.js]: :art:  This is extremely minor, but `pathToElm` to me could be read as `path to elm files`, rather than `path to elm binary`

File: lib/elm-test.js [hunk 13]
```diff
@@ -285,10 +281,11 @@ function runElmTest() {
...
-    var cmd = [pathToElmPackage, "install", "--yes"].concat(cmdArgs).join(" ");
-
-    child_process.execSync(cmd, { stdio: "inherit", cwd: Init.elmPackageDir });
+    var packagesToInstall = Init.init();
+    if (packagesToInstall.length > 0) {
+      var cmd = [pathToElm, "install"].concat(packagesToInstall).join(" ");
+      child_process.execSync(cmd, { stdio: "inherit", cwd: Init.elmPackageDir });
+    }
 
     process.exit(0);
```
Comment (eeue56) [inline lib/elm-test.js]: Should there be some feedback in this case?

File: lib/elm-test.js [hunk 14]
```diff
@@ -437,99 +408,128 @@ function compileAllTests(testFilePaths) {
...
+      cwd: generatedCodeDir
     });
+    var output = result.stderr.toString(); // TODO: hopefully the next Elm 0.19 beta will print the JSON to stdout instead
+    output = output.replace(/^elm:.*no version information available \(required by elm\)\n/, ''); // TODO: hopefully the next Elm 0.19 beta will not have this message (only occurs on linux) mixed in with the JSON output
+    var report = JSON.parse(output);
+    if (report.type === "error" && report.title === "NO INPUT") {
+      // all transtive dependencies are listed already
+    } else if (report.type === "error" && report.title === "MISSING DEPENDENCIES") {
+      var missingDeps = {};
+      report.message[1].string.replace(/"([^"]*)": "([^"]*)"/g, function(_, name, version) {
```
Comment (eeue56) [inline lib/elm-test.js]: this could probably do with a comment

File: lib/elm-test.js [hunk 15]
```diff
@@ -661,13 +661,17 @@ function generateAndRunTests(tests, filePathArgs, generatedSrc, getGlobs) {
       };
     }
 
-    var watchedSourcePaths = fs
-      .readJsonSync(path.join(originalDir, "elm-package.json"), "utf8")
-      ["source-directories"].map(resolveWatchPath(originalDir));
+    var watchedSourcePaths;
+    var elmJson = fs.readJsonSync(path.join(projectRootDir, "elm.json"), "utf8");
+    if (elmJson["type"] === "package") {
```
Comment (eeue56) [inline lib/elm-test.js]: It might be worth moving all this kind of logic to a function. So instead we have `var watchedSourcePaths = getWatchedSourcePaths(elmJson['type'])`;

File: lib/elm-test.js (hunk 12 above)
Comment (avh4) [reply lib/elm-test.js]: Renamed to `pathToElmBinary` in 5316b2c

File: lib/elm-test.js (hunk 13 above)
Comment (avh4) [reply lib/elm-test.js]: You mean if there are no packages to install?   I don't think it's necessary because `Init.init()` will also be printing other output.  But also, I didn't dig into the `elm-test init` code much; just enough to make the tests pass.

File: lib/elm-test.js (hunk 14 above)
Comment (avh4) [reply lib/elm-test.js]: Added / cleaned up in 7895327

File: lib/elm-test.js (hunk 15 above)
Comment (avh4) [reply lib/elm-test.js]: Cleaned up in 7a530fc

File: .gitignore (hunk 11 above)
Comment (avh4) [reply .gitignore]: Ah, I added this because there were some things I had to debug where elm-test was cd'd into the wrong folder, and it helped to see where elm-stuff folders appeared in places where they shouldn't have been.  (Like elm-test will no longer result in a `tests/elm-stuff` folder appearing.

File: tests/FuzzFailing.elm [hunk 16]
```diff
@@ -107,7 +107,7 @@ testOxfordify =
             , test "runs a Debug.crash on purpose" <|
                 \() ->
                     oxfordify "Everything is normal"
-                        |> Debug.crash "this test runs a Debug.crash on purpose!"
+                        |> Debug.todo "this test runs a Debug.crash on purpose!"
```
Comment (mgold) [inline tests/FuzzFailing.elm]: Let's also change the name in the string literal.

File: installer.js (hunk 1 above)
Comment (avh4) [reply installer.js]: ✅

File: lib/runner.js (hunk 4 above)
Comment (avh4) [reply lib/runner.js]: ✅

File: src/Test/Reporter/Console/Format.elm (hunk 5 above)
Comment (avh4) [reply src/Test/Reporter/Console/Format.elm]: Moved to checklist for after rc and before stable release

File: tests/ci.js (hunk 7 above)
Comment (avh4) [reply tests/ci.js]: -> yes, this is a bug in the compiler beta.  https://elmlang.slack.com/archives/C13L7S5GR/p1528239699000411

File: installer.js (hunk 8 above)
Comment (avh4) [reply installer.js]: ✅

File: installer.js (hunk 10 above)
Comment (avh4) [reply installer.js]: ✅

File: .travis.yml (hunk 9 above)
Comment (avh4) [reply .travis.yml]: Moved to the checklist for between rc and stable release.

File: tests/FuzzFailing.elm (hunk 16 above)
Comment (avh4) [reply tests/FuzzFailing.elm]: Fixed in 3553345

**Comment by avh4** [comment]
@rtfeldman @lukewestby can this branch get merged to master now that Elm 0.19 is released?

**Comment by avh4** [comment]
oh... I guess this is the thing you wanted me to look at fixing the tests for?

**Comment by avh4** [comment]
Everything is merge (though apparently a lot got rebased?) into the `0.19` branch.
